### ISSUE TITLE
Update SmtpClient Library to one that supports modern protocols

### ISSUE THREAD

**Issue by thabaum** [dnnsoftware/Dnn.Platform#3120]
This has been long called for in many of my threads and I think I finally collected the answer as frustration burns deep trying to work hours troubleshooting this for years now.  

New articles and research has come to light that shows SmtpClient used by Log4Net flat out shouldn't be used.

It is really a security risk because we have to dumb our servers down to allow DNN to send emails.

In todays modern world of security and best practices DNN CE is not up to speed.

In

https://github.com/dnnsoftware/Dnn.Platform/issues/3008

I reference:
https://docs.microsoft.com/en-us/dotnet/api/system.net.mail.smtpclient?view=netframework-4.8

which references:
https://github.com/jstedfast/MailKit

Which shows the following:
![image](https://user-images.githubusercontent.com/13577556/66263204-82b38780-e7a3-11e9-9758-aefcc1b284d5.png)

enough said...

MailKit is recommended so I recommend going with the most popular and supported flow.

https://github.com/jstedfast/MailKit

https://www.infoq.com/news/2017/04/MailKit-MimeKit-Official/

I suppose this is sort of old news as 2 years ago it seems Microsoft said to move to MailKit as it just simply works a lot better.

According to that last article link MailKit is the second open source library to be a replacement library with Newtonsofts JSON.NET being the first.

Simply put MailKit library replacing current library is the next step for the mail server issues to resolve.   I created a new environment and double checked every way possible to make it work

That article of 2017 was in april, it was DNN 9 in August of that year and ever since then and even prior to DNN 9 going all the way back to 5 I have had issues with SMTP not working the way I figured it should...  this really shines light on the subject.  This is when all my authentication systems went down I have not been able to use them since as well.  A major shift in security pushed things ahead faster than I think most open source projects can keep up with.   Thanks for any further support here!

**Comment by mitchelsellers** [comment]
Just to put a bit of thought on this, from my side.

1. SmtpClient is NOT marked Obsolete for .NET Framework Implementations and is still the supported and recommended pathway from Microsoft.  The documentation link you referenced is a bug reported by Microsoft https://github.com/dotnet/dotnet-api-docs/issues/2986

2.  I am all for modernization of the SMTP processes within DNN, in fact, I personally would like to see this switched to a provider implementation, or otherwise, that we can override.  As some may want to use MailKit, some may want to use SendGrid or otherwise and it would be incredibly helpful.

3.  From an open-source project perspective, every dependency we pull on an external library is a risk.  If we were to include MailKit, we would need to evaluate the risks.

I think at this point in time we need to identify the features missing.  We use Secure protocols from the DNN Processes today for many sites, but there are things that we can't do that we most likely should be able to.

**Comment by thabaum** [comment]
I totally agree and understand your views on this.  Mailkit is recommended by Microsoft from what I am reading and it is pretty flawless for the most part from what I have read.

But like you said it would be nice to have an ability to switch to the next best thing or whatever works as there could be some other one down the road.   SmtpClient is not oblsolet... but it is very limited from what I am reading and pretty much like you said Microsoft sees an open source library that hits the nail on the head for what needs to be done and has adopted it and made a point to switch to it in their documentation.

I do a search for .NET open source SMTP library and MailKit comes up first.   This is saying something pretty loud and clear to me that MailKit needs to be a part of whatever is going to be the solution for a start up just like using and including newtonsofts JSON.NET

I am up for anything that makes it so I can authenticate before sending mail at least.   I believe that the NTLM authentication setting is what is needed along with credentials for establishing that connection like in documentation would help cure that.

SSL setting isnt really needed as what you need is just a simple unsecured connection to the server and STARTTLS pushed to the maill server prior to authenticating and sending/receiving emails.

587 is the port that is best practice and for security should be default along with a STARTTLS connection on that port.  I may be wrong I have been studying this a long time but please correct me as I am trying to get my head wrapped around all of the latest changes to the protocols, ports and now library used to transmit messages via different mail servers.

In my Outlook you set the sender settings to the following and it works.  On this SMTP connector I do not have SSL enabled... I have port 465 with SSL enabled.  I can almost connect to that using DNN but it does not cut it for some reason.  But I can go on my outlook and change all the settings exactly how my mail server is setup for each port 25 465 and 587 everything works as expected.

Authentication prior to sending I believe can be fixed in current version but after reading that article I just added to my first post and other factors make it obsolete in my eyes.

**Comment by mitchelsellers** [comment]
SSL support for SMTP is still used by many servers, and removing support would be a huge breaking change.

Authentication before sending is working with DNN for thousands of sites, with various schemes including Basic & NTLM.

Just note that the Microsoft recommendation for MailKit DOES NOT apply to DNN as it is not a recommendation for .NET Framework.  

This is a pretty big change, but something that I too would like to see, but will require extensive documentation and testing prior to implementation.

**Comment by thabaum** [comment]
![image](https://user-images.githubusercontent.com/13577556/66272416-4a9d5a80-e81e-11e9-8fda-d990a018e90d.png)

The thing about MailKit is it is developed out and supported regularly, practically flawless and supports all of those things we are discussing.

I agree it is a pretty major undertaking but it is long overdue.   And making it so DNN can more easily switch to a new library would be nice but probably make the job even more tougher to make happen.

The thing about SSL supported by many servers... many servers do not support SSL so it is breaking not changing.  Which is why I call it a showstopper.  Anyone with money wont even touch DNN CE because this is bugged as soon as you try to setup your mail client you got problems unless the mail server is local or supports older SSL, which I still have issues connecting via SSL... EnableSSL is suppose to now use STARTTLS I believe I read that on the Log4Net. 

But even the current version should get fixed to allow authentication before sending.  This can be fixed now and I will try to work on it over the next couple weeks see if I can figure it out.   Maybe I can get a good enough understanding I can replace with MailKit and add all the features I need :)

**Comment by mitchelsellers** [comment]
Authentication is supported, and thousands of sites send email without a problem. mailKit isn’t  a silver bullet. But regardless we have this logged if we can get someone that wants to work on this we can plan for how to possibly implement a shift.

**Comment by thabaum** [comment]
Sounds good just trying to get the ball rolling.
I dont believe in silver bullets anymore with the internet and technology, but I can't find anything bad about MailKit currently.  Can I create a website or a kickstarter page for DNN CE so we can get some funds and get the ball rolling faster...  Can't we get donations collected and see them get put to use in a transparent way to make this open source project keep up with the times.   Or is our neck under the foot of Corporate DNN for something like that?

**Comment by mitchelsellers** [comment]
There are a lot of considerations here, and this issue isn’t necessarily the best place to discuss some of them. 

If you want feel free to email me.

**Comment by thabaum** [comment]
> Authentication is supported, and thousands of sites send email without a problem. 

Can you send email with authentication prior to sending?  I tried to put an issue up to see if this can be included in the current library but it was taken down.  It was a seperate issue from this issue as it seemed like this will take decades and I need something 3 years ago...

**Comment by mitchelsellers** [comment]
There is an NTLM setting in the SMTP settings, which does exactly what you want.  I can try to dig up some screenshots, but it will be a few days

**Comment by valadas** [comment]
@thabaum can you explain what is not working exactly, from the looks of it NTLM is supported as well as standard authenticatio and with or without SSL, not sure what this issue is other than a suggestion to change to another library which right now we have no need to....

> Can you send email with authentication prior to sending?
I do not understand what you mean by this sentence, some mail servers just accept sending without authentication (like local server that accepts only local connections or ip filters, etc.), some other servers do require authentication prior to sending out emails (in which case it is the basic authentication settings with or without SSL depending on what you need), ports can be specified in the server address too. Finally NTLM can also be used but in that case the worker needs to be authorized which is outside the Dnn scope to configure. 

Soooo the question is, what is not supported right now? Are you having issues with one of those methods?

**Comment by thabaum** [comment]
I found the issue.  SSL Required was enabled on my server.  I was able to verify in logs that a TLS connection was being created however I needed to turn SSL Required off and only have Allow Authentications from SSL/TLS connections only setting turned on.

SSL is now TLS, so the confusion was on my part thinking I needed SSL turned on the mail server when in fact I just need it turned on with DNN and the servers will communicate using TLS.

An "SSL certificate" is actually a "TLS Certificate".  So that issue was cured.

Now that I studied the libraries more I believe what may be missing is OAUTH and OAUTH2 support and asynchronous settings possibly unless the current library can handle these.  But other than that not as big of a deal now that I can confirm how the SMTP works with SSL enabled on DNN.  I finally cleared up some fog in my mind on that topic.  

I was having an IO issue trying to connect via SSL port but once I turned SSL Required off I could send mail on that port.  If anything maybe this will help the next guy like me.
